pkgs/misc/crust/default.nix

```diff
@@ -0,0 +1,154 @@
+###! Libre Power Management firmware for sunxi System On Chip ("SoC") for implementing system suspend/resume, and (onboards without a PMIC) soft poweroff/on. The package has to be built using (OpenRISC 1000, *not RISC-V*) compiler which is officially supported in upstream GCC starting with GCC 9.1.0.
+###!
+###! To build only the crust firmware derivation use:
+###!   $ nix-build -A pkgsCross.or1k.crustTeresA64
+###!
+###! When packaging for linux make sure that you have the following patches otherwise crust won't be able to cleanly share the clock controller and PMIC bus controller hardware: https://github.com/torvalds/linux/compare/master...crust-firmware:linux:crust
+###!
+###! If you are faced with issues while building the firmware then take a look at https://github.com/crust-firmware/crust#building-the-firmware
+
+{ stdenv
+, lib
+, fetchFromGitHub
+, fetchpatch
+, bison
+, flex
+, buildPackages
+}:
+
+###! The package declaration is heavily inspired by the U-Boot package to make maintenance easier, but should stay as an invidual declaration since it's processing has different requirements and including it in the U-Boot would bloat the definition and would be harder to understand and maintain
+
+let
+  defaultVersion = "unstable-2023-03-04";
+  defaultSrc = fetchFromGitHub {
+    owner = "crust-firmware";
+    repo = "crust";
+    rev = "c308a504853e7fdb47169796c9a832796410ece8";
+    sha256 = "AobVD3Jo/6s0cExHXpYYAqZv/gCplxLlDYVscSCIS6M=";
+  };
+  buildCrust = {
+    version ? "${defaultVersion}"
+  , src ? null
+  , filesToInstall
+  , installDir ? "$out"
+  , defconfig
+  , extraConfig ? ""
+  , extraPatches ? []
+  , extraMakeFlags ? []
+  , extraMeta ? {}
+  , ... } @ args: stdenv.mkDerivation ({
+    # Strip "_defconfig" suffix from 'defconfig' if it's present for clearer output
+    # builtins.match returns 'Null' if the string matches so we have to then check if it's Null
+    # FIXME-QA(Krey): I hate this, but am not aware of better way to write this
+    pname = if (builtins.isNull ((builtins.match "_defconfig$" defconfig)))
+      then "crust-${lib.strings.removeSuffix "_defconfig" defconfig}"
+      else "crust-${defconfig}";
```

Comment:
> I don't understand the intent here. -- @samueldr (https://github.com/NixOS/nixpkgs/pull/241360#discussion_r1253807260)

The intent is removing the `_defconfig` suffix for clearer nix output

> What are the situations where `_defconfig` would not be present? -- @samueldr (https://github.com/NixOS/nixpkgs/pull/241360#discussion_r1253807260)

When building non-standard configuration e.g. non-standard teres forks and i overall i like to make my code as flexible as possible to allow for future expansions.

> Also: FIXME-QA, what? if you meant to ask questions about this, I'm not aware you have. -- @samueldr (https://github.com/NixOS/nixpkgs/pull/241360#discussion_r1253807260)

Those are code indexing comments after self-review as i am not happy with the declaration of:

```nix
(builtins.isNull ((builtins.match "_defconfig$" defconfig)))
```

As i find it hard to read and it's doing `builtins.match` for regex suffix of `_defconfig$` which returns `Null` (would argue for adjusting the `buildins.match` returns) which then has to be processed with `builtins.isNull` to get a Boolean for `if` conditional.

I prefer to write code so that it's easily readable e.g.

```sh
# In case `$pname` ends with `_defconfig` -> set pname without _defconfig
case "$pname" in *"_defconfig") pname="${pname//_defconfig/}"; esac
```

So that the code can be read in plain english by those who understand it's syntax comparing to:

```nix
# if match of '_defconfig$' from defconfig is null then ...
if (builtins.isNull ((builtins.match "_defconfig$" defconfig))) ...
```

> Anyway, pick a lane and stick to it, either it always has _defconfig or never has it. Coding "defensively" in a declarative environment is not useful, and only adds noise. -- @samueldr (https://github.com/NixOS/nixpkgs/pull/241360#discussion_r1253807260)

It's not defensive coding it has a functional use as explained above.